Tools/clinic/clinic.py

```diff
@@ -5586,23 +5713,34 @@ def add_parameter(text: str) -> None:
 
         return docstring
 
-    def do_post_block_processing_cleanup(self) -> None:
+    def do_post_block_processing_cleanup(self, lineno: int) -> None:
         """
         Called when processing the block is done.
         """
         if not self.function:
             return
 
-        if self.keyword_only:
-            values = self.function.parameters.values()
-            if not values:
-                no_parameter_after_star = True
+        def check_remaining(
+                symbol: str,
+                condition: Callable[[Parameter], bool]
+        ) -> None:
+            assert isinstance(self.function, Function)
+
+            if values := self.function.parameters.values():
+                last_param = next(reversed(values))
+                no_param_after_symbol = condition(last_param)
             else:
-                last_parameter = next(reversed(list(values)))
-                no_parameter_after_star = last_parameter.kind != inspect.Parameter.KEYWORD_ONLY
-            if no_parameter_after_star:
-                fail(f"Function {self.function.name!r} specifies '*' "
-                     "without any parameters afterwards.")
+                no_param_after_symbol = True
```

Comment:
This line doesn't have any test coverage

Reply:
😱

Reply:
That's strange, it should be covered by `test_parameters_required_after_star`

Reply:
Ah, it is unreachable.

Reply:
Maybe this, then?

```diff
-
-            if values := self.function.parameters.values():
-                last_param = next(reversed(values))
-                no_param_after_symbol = condition(last_param)
-            else:
-                no_param_after_symbol = True
+            values = self.function.parameters.values()
+            assert values
+            last_param = next(reversed(values))
+            no_param_after_symbol = condition(last_param)
```

Reply:
```
/*[clinic input]
myfunc
    *
[clinic start generated output]*/
```
```
/*[clinic input]
myfunc
    * [from 3.14]
[clinic start generated output]*/
```

Reply:
> ```
> /*[clinic input]
> myfunc
>     *
> [clinic start generated output]*/
> ```
> 
> 
>     
>   
> 
> ```
> /*[clinic input]
> myfunc
>     * [from 3.14]
> [clinic start generated output]*/
> ```

They will still not trigger that branch, since `values` will always evaluate to true, even for this input.

Reply:
I added d2130abe8d2ffb250cf13585ea816cb724690e99 to address this.

Reply:
How is it?

In any case it is better to not touch it in this PR.

Reply:
Yes, I also find it confusing that `values` is not empty in that case. But I agree with Serhiy -- better not to touch it in this PR :)

Reply:
Well, I did touch it; I don't think it is worth it to revert the change.

> How is it?

The first item of `values()` is (almost) always a Module object. Even for input like this:

```c
/*[clinic input]
fn
    *
[clinic start generated code]*/
```

Reply:
I mean, the old code supports empty `values`, your code keeps the support. If it is no longer needed (it is not obvious), leave it to a separate issue. This PR already have too many comments.

Reply:
Ok, I'll revert it.

Reply:
Oh, I misunderstood Serhiy's remarks above; apologies @serhiy-storchaka

Reply:
FTR, I'm not a massive fan of introducing new code that we know to be untested and unreachable. But, it's only one line of code, so it doesn't matter too much either way; happy to go with Serhiy's preference here :)

Reply:
It is not a new code that we know to be untested. It is a rewriting of the old code that preserves the logic. The corresponding branch in the old code obviously was untested too. It is not a problem of _this_ PR.

Reply:
> It is not a new code that we know to be untested. It is a rewriting of the old code that preserves the logic. The corresponding branch in the old code obviously was untested too. It is not a problem of _this_ PR.

Good point, I didn't look closely enough at the diff!